community/overrides/user.py

```diff
@@ -15,36 +15,33 @@ def validate(self):
     def validate_username_characters(self):
         if self.is_new():
 
+            if not self.username:
+                self.username = self.get_username_from_first_name()
+
             if self.username.find(" "):
                 self.username.replace(" ", "")
 
             if not re.match("^[A-Za-z0-9_]*$", self.username):
                 self.username = self.remove_illegal_characters()
```

Comment:
If the username is provided by the user, then you should given an error instead of silently removing the invalid characters. 

If the username is generated by the system, they it could have removed the invalid characters at that time and there no need for this check.

Reply:
The system generates the username using the first_name. If during signup a user adds any illegal character to their name, it reflects in the username too. So this check is needed.

Reply:
> The system generates the username using the first_name. If during signup a user adds any illegal character to their name, it reflects in the username too. So this check is needed.

I'm not debating whether or not the check is needed. I'm saying that the check is in the wrong place. 

If a function is taking the responsibiltiy of generating a username, then it should generate a valid username.

Reply:
It would be. The only check missing was for username shorter than 4 characters. I have added a condition for the same.